packages/xforms-engine/test/parse/xpath/dependency-analysis.test.ts

```diff
@@ -0,0 +1,337 @@
+			expected: ['/data/foo/bar/quux'],
+		},
+
+		{
+			description: 'Multiple parent steps resolved',
+
+			contextNodeset: '/data/foo/bar/bat/quux/zig/zag',
+			expression:
+				'concat(../sibling, ../../parent-sibling, ../../../impertinent/../grandparent-sibling)',
+			expected: [
+				'/data/foo/bar/bat/quux/zig/sibling',
+				'/data/foo/bar/bat/quux/parent-sibling',
+				'/data/foo/bar/bat/grandparent-sibling',
+			],
+		},
+
+		{
+			description: 'Absolute context; self reference (direct or redundant) resolved to context',
+
+			contextNodeset: '/data/foo',
+			expression: '. = 1 or ./. = 2',
+			expected: ['/data/foo'],
+		},
+
+		{
+			description: 'Relative context; self reference (direct or redundant) resolved to context',
+
+			contextNodeset: 'foo',
+			expression: '. != false() and ./. != false()',
+			expected: ['foo'],
+		},
+
+		{
+			description: 'No context; self reference preserved',
+
+			contextNodeset: null,
+			expression: '. > now()',
+			expected: ['.'],
+		},
+
+		{
+			description: 'No context; redundant self reference simplified',
+
+			contextNodeset: null,
+			expression: './. > now()',
+			expected: ['.'],
+		},
+
+		{
+			description: 'Unabbreviated self::node() and self::* treated as .',
+
+			contextNodeset: '/data/foo',
+			expression: 'self::node()/bar | self::node()/quux',
+			expected: ['/data/foo/bar', '/data/foo/quux'],
+		},
+
+		{
+			description: 'Unabbreviated parent::node() and parent::* treated as ..',
+
+			contextNodeset: '/data/foo',
+			expression: 'parent::node()/bar and parent::*/quux',
+			expected: ['/data/bar', '/data/quux'],
+		},
+
+		{
+			description: 'self::node() context, parent::node()/* collapsed like ./../*',
+
+			contextNodeset: 'self::node()',
+			expression: 'parent::node()/bar',
+			expected: ['../bar'],
+		},
+
+		{
+			description: 'self::* context, parent::*/* collapsed like ./../*',
+
+			contextNodeset: 'self::*',
+			expression: 'parent::*/bar',
+			expected: ['../bar'],
+		},
+
+		{
+			description: '// special cases',
+
+			contextNodeset: null,
+			expression: '//. | //./foo | //.. | //../foo',
+			expected: ['//.', '//foo', '//..'],
+		},
+	])('$description', ({ contextNodeset, expression, expected }) => {
+		it(`resolves nodeset dependencies in expression ${JSON.stringify(expression)}, with context ${JSON.stringify(contextNodeset)}, producing nodesets: ${JSON.stringify(expected)}`, () => {
+			const actual = resolveDependencyNodesets(contextNodeset, expression);
+
+			expect(actual).toEqual(expected);
+		});
+	});
+
+	// TODO! This was explored but it introduced a variety of complexities beyond
+	// an already fairly large chunk of work on this pass. Note also that
+	// parentheses can have surprising effects on a path expression's semantics!
+	// (For instance: see note on predicates and ordering in
+	// https://www.w3.org/TR/1999/REC-xpath-19991116/#node-sets).
+	it.fails('resolves parenthesized portions of sub-expressions', () => {
```

Comment:
Feels low priority but good to have a test for!

Reply:
Probably something we should get some data on. I don't know how much parenthesized Exprs are used in real world.

I do know that I'd want more coverage of it in `xpath` as well, particularly around the note linked in this comment.